embedded/document/engine.go

```diff
@@ -151,539 +52,780 @@ func NewEngine(store *store.ImmuStore, opts *sql.Options) (*Engine, error) {
-		if err != nil {
-			return nil, 0, 0, err
+	return txID, docIDs, nil
+}
+
+func (e *Engine) generateRowSpecForDocument(table *sql.Table, doc *structpb.Struct) (*sql.RowSpec, error) {
+	idFieldName := docIDFieldName(table)
+
+	values := make([]sql.ValueExp, len(table.Cols()))
+
+	for i, col := range table.Cols() {
+		if col.Name() == idFieldName {
+			docID, err := NewDocumentIDFromHexEncodedString(doc.Fields[col.Name()].GetStringValue())
+			if err != nil {
+				return nil, err
+			}
+
+			values[i] = sql.NewBlob(docID[:])
+			continue
 		}
 
-		err = e.sqlEngine.GetStore().WaitForIndexingUpto(ctx, txID)
-		if err != nil {
-			return docID, txID, 0, nil
+		if col.Name() == DocumentBLOBField {
+			bs, err := json.Marshal(doc)
+			if err != nil {
+				return nil, err
+			}
+
+			values[i] = sql.NewBlob(bs)
+			continue
 		}
 
-		docAudit, err := e.getEncodedDocumentAudit(searchKey, 0, false)
-		if err != nil {
-			return nil, 0, 0, err
+		if rval, ok := doc.Fields[col.Name()]; ok {
+			val, err := structValueToSqlValue(col.Type(), rval)
+			if err != nil {
+				return nil, err
+			}
+			values[i] = val
+		} else {
+			values[i] = &sql.NullValue{}
+		}
+	}
+
+	return sql.NewRowSpec(values), nil
+}
+
+func (e *Engine) UpdateDocument(ctx context.Context, collectionName string, query *protomodel.Query, doc *structpb.Struct) (txID uint64, docID DocumentID, rev uint64, err error) {
+	if doc == nil {
+		return 0, nil, 0, ErrIllegalArguments
+	}
+
+	sqlTx, err := e.sqlEngine.NewTx(ctx, sql.DefaultTxOptions())
+	if err != nil {
+		return 0, nil, 0, mayTranslateError(err)
+	}
+	defer sqlTx.Cancel()
+
+	table, err := getTableForCollection(sqlTx, collectionName)
+	if err != nil {
+		return 0, nil, 0, err
+	}
+
+	idFieldName := docIDFieldName(table)
+
+	provisionedDocID, docIDProvisioned := doc.Fields[idFieldName]
+	if docIDProvisioned {
+		// inject id comparisson into query
+		idFieldComparissonExp := &protomodel.QueryExpression{
+			FieldComparisons: []*protomodel.FieldComparison{
+				{
+					Field:    idFieldName,
+					Operator: protomodel.ComparisonOperator_EQ,
+					Value:    provisionedDocID,
+				},
+			},
+		}
+
+		if query == nil || len(query.Expressions) == 0 {
+			query = &protomodel.Query{
+				Expressions: []*protomodel.QueryExpression{idFieldComparissonExp},
+			}
+		} else {
+			// id comparisson as a first expression might result in faster comparisson
+			query.Expressions = append([]*protomodel.QueryExpression{idFieldComparissonExp}, query.Expressions...)
 		}
+	}
+
+	queryCondition, err := e.generateSQLExpression(ctx, query, table)
+	if err != nil {
+		return 0, nil, 0, err
+	}
+
+	queryStmt := sql.NewSelectStmt(
+		[]sql.Selector{sql.NewColSelector(collectionName, idFieldName)},
+		collectionName,
+		queryCondition,
+		sql.NewInteger(1),
+		nil,
+	)
+
+	r, err := e.sqlEngine.QueryPreparedStmt(ctx, sqlTx, queryStmt, nil)
+	if err != nil {
+		return 0, nil, 0, mayTranslateError(err)
+	}
+	defer r.Close()
+
+	row, err := r.Read(ctx)
```

Comment:
you may want to check the length of the result, if it is one or not

Reply:
I know there is a limit in the query, I thought it is upto the user to select one document based on their query and not restrict it to the first found through the query

Reply:
it will update one document, returning the id of the updated one. delete document having similar behaviour